shadowsocks/stream_test.go

```diff
@@ -411,3 +411,31 @@ func TestLazyWriteConcurrentFlush(t *testing.T) {
 		t.Errorf("Wrong final content: %v", decrypted)
 	}
 }
+
+type nullIO struct{}
+
+func (n *nullIO) Write(b []byte) (int, error) {
+	return len(b), nil
+}
+
+func (r *nullIO) Read(b []byte) (int, error) {
+	return len(b), nil
+}
+
+// Microbenchmark for the performance of Shadowsocks TCP encryption.
+func BenchmarkWriter(b *testing.B) {
+	b.StopTimer()
+	b.ResetTimer()
+
+	cipher := newTestCipher(b)
+	writer := NewShadowsocksWriter(new(nullIO), cipher)
```

Comment:
This is not writing anything, so I think you are not even testing the encryption. Are you?

Reply:
The argument here is the ciphertext output.  `io.CopyN` should be writing `b.N` zeros into `writer` further down.